src/ts_catalog/continuous_aggs_watermark.c

```diff
@@ -0,0 +1,401 @@
+cagg_watermark_update_scan_internal(TupleInfo *ti, void *data)
+{
+	WatermarkUpdate *watermark_update = data;
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	Form_continuous_aggs_watermark form = (Form_continuous_aggs_watermark) GETSTRUCT(tuple);
+
+	if (watermark_update->watermark > form->watermark || watermark_update->force_update)
+	{
+		HeapTuple new_tuple = heap_copytuple(tuple);
+		form = (Form_continuous_aggs_watermark) GETSTRUCT(new_tuple);
+		form->watermark = watermark_update->watermark;
+		ts_catalog_update(ti->scanrel, new_tuple);
+		heap_freetuple(new_tuple);
+	}
+	else
+	{
+		elog(DEBUG1,
+			 "hypertable %d existing watermark >= new watermark " INT64_FORMAT " " INT64_FORMAT,
+			 form->mat_hypertable_id,
+			 form->watermark,
+			 watermark_update->watermark);
+		watermark_update->watermark = form->watermark;
+	}
+
+	if (should_free)
+		heap_freetuple(tuple);
+
+	return SCAN_DONE;
+}
+
+static void
+cagg_watermark_update_internal(int32 mat_hypertable_id, int64 new_watermark, bool force_update)
+{
+	bool watermark_updated;
+	ScanKeyData scankey[1];
+	WatermarkUpdate data = { .watermark = new_watermark, .force_update = force_update };
+
+	ScanKeyInit(&scankey[0],
+				Anum_continuous_aggs_watermark_mat_hypertable_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(mat_hypertable_id));
+
+	watermark_updated = ts_catalog_scan_one(CONTINUOUS_AGGS_WATERMARK /*=table*/,
+											CONTINUOUS_AGGS_WATERMARK_PKEY /*=indexid*/,
+											scankey /*=scankey*/,
+											1 /*=num_keys*/,
+											cagg_watermark_update_scan_internal /*=tuple_found*/,
+											RowExclusiveLock /*=lockmode*/,
+											CONTINUOUS_AGGS_WATERMARK_TABLE_NAME /*=table_name*/,
+											&data /*=data*/);
+
+	if (!watermark_updated)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("watermark not defined for continuous aggregate: %d", mat_hypertable_id)));
+	}
+}
+
+TSDLLEXPORT void
+ts_cagg_watermark_update(Hypertable *mat_ht, int64 watermark, bool watermark_isnull,
+						 bool force_update)
+{
+	ContinuousAgg *cagg = ts_continuous_agg_find_by_mat_hypertable_id(mat_ht->fd.id);
+
+	if (NULL == cagg)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid materialized hypertable ID: %d", mat_ht->fd.id)));
+
+	watermark = cagg_compute_watermark(cagg, watermark, watermark_isnull);
+	cagg_watermark_update_internal(mat_ht->fd.id, watermark, force_update);
+
+	return;
+}
+
+TSDLLEXPORT void
+ts_cagg_watermark_delete_by_mat_hypertable_id(int32 mat_hypertable_id)
+{
+	ScanIterator iterator =
+		ts_scan_iterator_create(CONTINUOUS_AGGS_WATERMARK, RowExclusiveLock, CurrentMemoryContext);
+
+	cagg_watermark_init_scan_by_mat_hypertable_id(&iterator, mat_hypertable_id);
+
+	ts_scanner_foreach(&iterator)
+	{
+		TupleInfo *ti = ts_scan_iterator_tuple_info(&iterator);
+		ts_catalog_delete_tid(ti->scanrel, ts_scanner_get_tuple_tid(ti));
+	}
```

Comment:
Nit. The iterator will automatically be closed if there are no more tuples, so this is just for good measure.

```suggestion
	}
    	ts_scan_iterator_close(&iterator);
```

Reply:
Fixed!